### PR TITLE
Improve GOTO programs high-level documentation

### PR DESCRIPTION
**Pull request by smowton**
* Clarify that some things described as 'instrumentation' previously were actually part
  of the normal goto-convert process
* Remove duplicate exposition of the `goto_functionst` data structure
* Describe the typical post-goto-convert transformations briefly
* Describe lazy-goto-modelt

### CODE REVIEW

File: src/goto-programs/README.md [hunk 1]
```diff
@@ -241,39 +246,10 @@ void foo(){}"];
...
-
-This stage applies several transformations to the goto-programs from the
-previous stage:
-
-* The diagram in the previous stage showed a goto_programt with four
-  instructions, but real programs usually yield instruction lists that
-  are littered with \ref code_skipt "skip" statements. The
-  instrumentation stage removes the majority of these.
+The `codet` representation of method bodies that is stored in the symbol table
+resembles C code quite strongly: it uses lexical variable scoping and has
```
Comment (tautschnig) [inline src/goto-programs/README.md:250]: Should this some how refer to the parse tree, saying that `codet` is the type-checked version of the parse tree and thus uses lexical ...

File: src/goto-programs/README.md [hunk 2]
```diff
@@ -322,14 +298,63 @@ previous stage:
...
+
+Lazy conversion is coordinated by `lazy_goto_modelt`. Early source code parsing
+proceeds as usual, but while the same symbol table symbols are created as using
+`initialize_goto_model`, method bodies in the symbol table are not populated.
+`lazy_goto_modelt::get_goto_function` then allows each function to be converted
+on demand, passing individually through symbol-table `codet` production, GOTO
+conversion and driver-program transformation.
+
+`goto_symext` is able to use such a lazy GOTO model in place of the usual
+`goto_modelt`, enabling it to fault GOTO programs in on demand, and thus avoid
```
Comment (tautschnig) [inline src/goto-programs/README.md]: With English not being my first language, the use of "fault GOTO programs in" required me to think more than I would expect at this point. Can we find a different way of expressing the on-demand conversion requests?

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0d71f9b0607119655b0585b2b4d12bf49b7e7d02).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91196923

File: src/goto-programs/README.md [hunk 3]
```diff
@@ -322,14 +298,63 @@ previous stage:
...
+of these can be found in `cbmc_parse_optionst`, `goto_analyzer_parse_optionst`
+and `goto_instrument_parse_optionst` among others.
+
+\section section-lazy-conversion Lazy GOTO Conversion
+
+The conventional source-code-to-GOTO-program pipeline is horizontally
+structured: all source-file information is converted into a symbol table, then
+each symbol-table method is converted into a GOTO program, then each transform
+is run over the whole GOTO-model in turn. This is fine when the GOTO model
+consumer will use most or all of the results of conversion, but it not then
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: if not

File: src/goto-programs/README.md [hunk 4]
```diff
@@ -322,14 +298,63 @@ previous stage:
...
+`lazy_goto_modelt::get_goto_function` then allows each function to be converted
+on demand, passing individually through symbol-table `codet` production, GOTO
+conversion and driver-program transformation.
+
+`goto_symext` is able to use such a lazy GOTO model in place of the usual
+`goto_modelt`, enabling it to fault GOTO programs in on demand, and thus avoid
+spending time and memory generating GOTO programs that symex determines cannot
+in fact be reached. This is particularly useful when the source program contains
+many indirect calls (virtual function calls, or calls via function pointer), as
+the `remove_virtual_funtions` and `remove_function_pointers` passes treat these
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: remove_virtual_functions

**Review by peterschrammel — APPROVED** [review]

File: src/goto-programs/README.md (hunk 1 above)
Comment (smowton) [reply src/goto-programs/README.md:250]: That's only true for C though -- Java and other code doesn't necessarily follow that pattern

File: src/goto-programs/README.md (hunk 2 above)
Comment (smowton) [reply src/goto-programs/README.md]: changed fault -> create

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f1f2da1b17b3269e9e8f4e13a1ef9d097883fcfe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91204347
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 645c11cceda27c56b5471ff8e193701a9e3b9bc2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91204477
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 008c58079e5be8eb7a08e57219afe64b51ac0f47).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91222793

**Pull request merged**